### ISSUE TITLE
Memory leak on nomad servers

### ISSUE THREAD

**Issue by D4GGe** [hashicorp/nomad#14842]
### Nomad version
`Nomad v1.3.1 (bc49012743edc0e2adb565e0fbcbc36c184f8ce4+CHANGES)`

### Operating system and Environment details

**3 nomad servers**

```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal

```
```
8 cores
32 GiB ram
```

**~10 large nomad clients** 

```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal

```
```
96 Cores
384 GiB ram
8 x nvidia t4
```
No nvidia plugin installed

**~20 nomad clients of various sizes**

### Issue
**Memory leak on nomad servers**

We have a memory leak on our nomad servers, the cluster have around 30 clients and the memory leak shows when we run a lot of of short lived jobs (parameterized job ~100 new allocs per hour), but seams to build up even if we lower the load just slower.

The only way for us to fix this is to run an `nomad system gc` on one of the servers. Otherwise it uses up all memory on the server until the server needs a reboot to be usable again. 
When this happens the only way to start nomad without nomad using 100% memory is to remove the `data_dir` and start over.

Right now in our production we run automatically `nomad system gc`  each time memory goes over 30% and this is how it looks.

![screenshot](https://user-images.githubusercontent.com/8104271/194497272-deb1a277-0629-43df-bcc8-279faf015511.png)

**Are there any way of analyzing what actually take up this amount of memory?
Can I analyze the `data_dir` in any meaningful way?
This log show up ever now and then can it have anything to do with it?**

```
2022-10-07T03:53:13.369Z [ERROR] worker: error waiting for Raft index: worker_id=599a4aeb-925e-0764-362f-844469f18e71 error="timed out after 5s waiting for index=23307547" index=23307547
```

### Configuration
***Server***
```HCL
data_dir = "/var/lib/nomad"
bind_addr = "0.0.0.0"
region = "{{region}}"
datacenter = "dc-master"
advertise {
	rpc = "{{ipify_public_ip}}"
}
server {
  enabled = true
  bootstrap_expect = 3
  encrypt = "{{gossipkey}}"
   server_join {
    retry_join = ["{{ipa}}", "{{ipb}}", "{{ipc}}"]
    retry_max = 0
    retry_interval = "15s"
  }
   default_scheduler_config {
    scheduler_algorithm = "spread"

    preemption_config {
      batch_scheduler_enabled   = false
      system_scheduler_enabled  = true
      service_scheduler_enabled = false
    }
  }
}

client {
  enabled = false
}


tls {
  http = false
  rpc  = true

  ca_file   = "--"
  cert_file = "--"
  key_file  = "--"

  verify_server_hostname = true
  verify_https_client    = true
}
```

***Clients***
```HCL
data_dir = "/var/lib/nomad"
bind_addr = "0.0.0.0"
region = "{{region}}"
datacenter = "dc-edge-node"

plugin "raw_exec" {
  config {
    enabled = true
  }
}



plugin "docker" {
  config {
    volumes {
      enabled = true
    }
    gc {
      image       = true
      dangling_containers {
        period         = "12h"
      }
    }
  }
}

client {
  cpu_total_compute = {{ 1170000 if num_gpus else 0 }}
  servers = ["{{ipa}}", "{{ipb}}", "{{ipc}}"]
  enabled = true
  node_class = "{{nodeclass}}"
  ip_resolver_endpoint = "{{ipresolverendpoint}}"
  meta {
    gpu.numcores = {{num_gpus}}
    features = "{{features}}"
    region = "{{nodeclass}}"
    service = "{{service}}"
    edge_node.environment = "{{edgenodeenvironment}}"
    edge_node.datacenter_cache_endpoint = "{{datacentercacheendpoint}}"
  }
  options {
    gc_interval = "1m",
    gc_max_allocs = 50,
    docker.cleanup.image = "false"
  }
}
tls {
  http = true
  rpc  = true

  ca_file   = "--"
  cert_file = "--"
  key_file  = "--"

  verify_server_hostname = true
  verify_https_client    = true
}

```

### Job file (if appropriate)
We have many jobs running but this is the main one used with Levant
[job.hcl](https://github.com/hashicorp/nomad/files/9732274/job.txt)

### Nomad Server logs (if appropriate)
This is some sample logs from the server

[server-logs.txt](https://github.com/hashicorp/nomad/files/9732286/server-logs.txt)

**Comment by jrasell** [comment]
Hi @D4GGe and thanks for raising this issue. I don't believe you are experiencing a memory leak on the servers because, as you mention, running the Nomad internal GC process reduces the memory overhead down. It seems like this is therefore normal behaviour on Nomad's part and I will try to explain in a little more detail below what is happening along with some configuration options that might help.

> run a lot of of short lived jobs (parameterized job ~100 new allocs per hour)

Each job registration results in a number of objects (jobspec, evaluation, allocation, deployment, etc...) being created and subsequently stored within the Nomad state store. This state store is held stored in memory and replicated to each server via Raft from the leader. I would therefore expect Nomads memory usage to grow as more jobs and allocations are added to the cluster.

> fix this is to run an `nomad system gc` on one of the servers

The Nomad leader runs a number of internal routines to clean up its state store at regular intervals. This interval usually defaults to 5 minutes. At each interval the leader will identify state objects such as jobs, evaluations, allocations, and deployments which have reached a terminal state, are no longer needed as reference by other objects, and have been in this terminal state for a configurable threshold time.

The `nomad system gc` command forces a run of all internal garbage collection processes and bypasses the threshold time for deletion. This results in any object that is considered terminal and collectable being deleted from state no matter how old it is.

The GC intervals and thresholds can be modified via the [server configuration block](https://developer.hashicorp.com/nomad/docs/configuration/server) to be more aggressive to account for high job turnover. There are a number of options; the [job_gc_interval](https://developer.hashicorp.com/nomad/docs/configuration/server#job_gc_interval) and [job_gc_threshold](https://developer.hashicorp.com/nomad/docs/configuration/server#job_gc_threshold) would be good places to start.

Hopefully this explanation makes sense and is useful. Please let me know if you have any follow up questions, otherwise I will close this issue after a few days.

**Comment by D4GGe** [comment]
Thank you for really good explanation!
I will experiment with [job_gc_interval](https://developer.hashicorp.com/nomad/docs/configuration/server#job_gc_interval) and [job_gc_threshold](https://developer.hashicorp.com/nomad/docs/configuration/server#job_gc_threshold) and come back!

Would it also help to increase numbers of servers and or increase the performance of them.
How much performance is common to need to have at the nomad servers in a case like ours?

**Comment by jrasell** [comment]
> Would it also help to increase numbers of servers and or increase the performance of them.

I don't believe increasing the number of servers will have any effect as each server stores a replica of the state store in memory. It has the potential to make situations worse, as the data needs to be replicate more times from the leader. Depending on how tuning the GC parameters goes, it could be beneficial to increase the available RAM for the servers, but I am confident the GC modifications are the correct solution.

> How much performance is common to need to have at the nomad servers in a case like ours?

We don't have exact figure or data and each environment is different. That being said, I have certainly seen this pattern before in clusters which have high turnover rates of batch workloads. As with above, I would expect the GC tuning to help make this a situation where the current specs are enough.

**Comment by D4GGe** [comment]
Hi I changed the parameters to this 

```HCL
  job_gc_interval = "30s" // defult 5m
  job_gc_threshold = "30m"  // defult 4h
  eval_gc_threshold = "15m" // defult 1h
```
And memory consumption seams to going up way slower
But we can still see a memory curve that seams to not stop going up this is over 3 days
![image](https://user-images.githubusercontent.com/8104271/194863083-c9725419-5702-4a3c-82b7-088b2ab3e335.png)

I know 12% isnt that high but im afraid it wont stop untill I run a manual `nomad system gc`
Do recommend to test with an even lower `job_gc_threshold` or are there other parameters I should play around with?

I just think it kind of weird that the the jobs take up more and more memory over 3 days when they are suppose to be collected after 30m

Also the plan is for the load to go up meaning higher turn over of jobs (in the 1000s per hour) is this at all sustainable with nomad?

**Comment by tgross** [comment]
Fixed in #15097 which will ship in Nomad 1.5.0 (with backports)
